src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,258 @@
+    return "iso8859-1";
+  case Encoding::UTF16LE:
+    return "utf16le";
+  case Encoding::UTF8:
+    return "utf-8";
+  default:
+    KJ_UNREACHABLE;
+  }
+}
+
+typedef kj::Maybe<kj::Array<kj::byte>> (*TranscodeImpl)(kj::ArrayPtr<kj::byte> source,
+                                                        Encoding fromEncoding, Encoding toEncoding);
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeDefault(kj::ArrayPtr<kj::byte> source,
+                                                Encoding fromEncoding, Encoding toEncoding) {
+  Converter to(toEncoding);
+  auto substitude = kj::str(kj::repeat('?', to.minCharSize()));
+  to.setSubstitudeChars(substitude);
+  Converter from(fromEncoding);
+
+  size_t limit = source.size() + to.maxCharSize();
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(limit <= ISOLATE_LIMIT, Error, "Source buffer is too large to transcode");
+  auto out = kj::heapArray<kj::byte>(limit);
+  char* target = out.asChars().begin();
+  const char* source_ = source.asChars().begin();
+  UErrorCode status{};
+  ucnv_convertEx(to.conv(), from.conv(), &target, target + limit, &source_, source_ + source.size(),
+                 nullptr, nullptr, nullptr, nullptr, true, true, &status);
+  if (U_SUCCESS(status)) {
+    return out.slice(0, target - out.asChars().begin()).attach(kj::mv(out));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeToUCS2(kj::ArrayPtr<kj::byte> source, Encoding fromEncoding,
+                                               Encoding toEncoding) {
+  Converter from(fromEncoding);
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(source.size() <= ISOLATE_LIMIT, Error, "Source buffer is too large to transcode");
+
+  auto destbuf = kj::heapArray<UChar>(source.size());
+  auto actual_length = simdutf::convert_latin1_to_utf16(source.asChars().begin(), source.size(), destbuf.begin());
+
+  // simdutf returns 0 for invalid value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.slice(0, actual_length).asBytes().attach(kj::mv(destbuf));
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeFromUCS2(kj::ArrayPtr<kj::byte> source,
+                                                 Encoding fromEncoding, Encoding toEncoding) {
+  Converter to(toEncoding);
+  auto substitude = kj::str(kj::repeat('?', to.minCharSize()));
+  to.setSubstitudeChars(substitude);
+
+  const size_t length_in_chars = source.size() * sizeof(UChar);
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(length_in_chars <= ISOLATE_LIMIT, Error, "Buffer is too large to transcode");
+
+  // TODO: Optimization opportunity: Remove sourcebuf allocation by calculating length beforehand.
+  KJ_STACK_ARRAY(UChar, sourcebuf, source.size(), 1024, 536870888);
+  memcpy(sourcebuf.begin(), sourcebuf.begin(), length_in_chars);
```

Comment:
Something doesn't quite make sense here.
1. You're copying from sourcebuf.begin() to sourcebuf.begin() which is a nullop?
2. You have allocated sourcebuf to be source.size() but because of the error in calculation of length_in_chars you are copying twice that amount from sourcebuf onto itself thus both reading out of bounds and writing out of bounds

Reply:
ha, good catch. My eyes completely overlooked that one

Reply:
yeah, we don't need the sourcebuf at all. I've updated the implementation.